homeassistant/components/sensor/imap_email_content.py

```diff
@@ -0,0 +1,249 @@
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the number of unread emails."""
+        return self._message
+
+    @property
+    def state_attributes(self):
+        """Return other state attributes for the message."""
+        return self._state_attributes
+
+    def render_template(self, email_message):
+        """Render the message template."""
+        variables = {
+            ATTR_FROM: EmailContentSensor.get_msg_sender(email_message),
+            ATTR_SUBJECT: EmailContentSensor.get_msg_subject(email_message),
+            ATTR_DATE: email_message['Date'],
+            ATTR_BODY: EmailContentSensor.get_msg_text(email_message)
+        }
+        return template.render(self.hass, self._value_template, variables)
+
+    def sender_allowed(self, email_message):
+        """Check if the sender is in the allowed senders list."""
+        return EmailContentSensor.get_msg_sender(email_message).upper() in (
+            sender.upper() for sender in self._allowed_senders)
+
+    @staticmethod
+    def get_msg_sender(email_message):
+        """Get the parsed message sender from the email."""
+        return str(email.utils.parseaddr(email_message['From'])[1])
+
+    @staticmethod
+    def get_msg_subject(email_message):
+        """Decode the message subject."""
+        decoded_header = email.header.decode_header(email_message['Subject'])
+        header = email.header.make_header(decoded_header)
+        return str(header)
+
+    @staticmethod
+    def get_msg_text(email_message):
+        """
+        Get the message text from the email.
+
+        Will look for text/plain or use text/html if not found.
+        """
+        message_text = None
+        message_html = None
+        message_untyped_text = None
+
+        for part in email_message.walk():
+            if part.get_content_type() == 'text/plain':
+                if message_text is None:
+                    message_text = part.get_payload()
+            elif part.get_content_type() == 'text/html':
+                if message_html is None:
+                    message_html = part.get_payload()
+            elif part.get_content_type().startswith('text'):
+                if message_untyped_text is None:
+                    message_untyped_text = part.get_payload()
+
+        if message_text is not None:
+            return message_text
+
+        if message_html is not None:
+            return message_html
+
+        if message_untyped_text is not None:
+            return message_untyped_text
+
+        return email_message.get_payload()
+
+    def update(self):
+        """Read emails and publish state change."""
+        while True:
+            email_message = self._email_reader.read_next()
+
+            if email_message is None:
+                break
+
+            if self.sender_allowed(email_message):
```

Comment:
Wouldn't it be better to first figure out what the latest relevant message is, and then set the data based on that? Now it can happen that you do a lot of work that will be overridden in the next iteration.